### PR TITLE
Fix crash in `no-member` check for C extensions

### LINKED ISSUE #6221: 'str' object has no attribute 'parent'
### Bug description

**Note**: Crash appeared several times in different files.  Including all of pylint-crash-2022-04-07-07.txt made the body text of this issue too big.

pylint crashed with a ``AttributeError`` and with the following stacktrace:
```
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\lint\pylinter.py", line 1124, in _check_files
    self._check_file(get_ast, check_astroid_module, file)
  File "c:\leo.repo\ekr-pylint\pylint\lint\pylinter.py", line 1159, in _check_file
    check_astroid_module(ast_node)
  File "c:\leo.repo\ekr-pylint\pylint\lint\pylinter.py", line 1311, in check_astroid_module
    retval = self._check_astroid_module(
  File "c:\leo.repo\ekr-pylint\pylint\lint\pylinter.py", line 1358, in _check_astroid_module
    walker.walk(node)
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 76, in walk
    self.walk(child)
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 76, in walk
    self.walk(child)
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 76, in walk
    self.walk(child)
  [Previous line repeated 3 more times]
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
```

**Note** Several more similar crashes appear in the crash report.

### Configuration

_No response_

### Command used

```shell
A script containing the following:

is_win = sys.platform.startswith('win')
args =  ','.join([f"'--rcfile={rc_fn}'", f"'{fn}'"])
if is_win:
    args = args.replace('\\','\\\\')
command = f'{sys.executable} -c "from pylint import lint; args=[{args}]; lint.Run(args)"'
if not is_win:
    command = shlex.split(command)
proc = subprocess.Popen(command, shell=False)
proc.communicate()
```

### Pylint output

```shell
C:\leo.repo\leo-editor>python -m pylint-leo -a

C:\leo.repo\leo-editor>echo off

pylint-leo: skip C:/leo.repo/leo-editor/leo/plugins/freewin.py
pylint: 254 files
................************* Module leo.commands.killBufferCommands
leo\commands\killBufferCommands.py:334:16: C2801: Unnecessarily calls dunder method __next__. Use next built-in function. (unnecessary-dunder-call)
........Exception on node <Attribute.connect l.98 at 0x21c63728a60> in file 'C:\leo.repo\leo-editor\leo\core\leoBackground.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.core.leoBackground
leo\core\leoBackground.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\core\leoBackground.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-06.txt'. (fatal)
....************* Module leo.core.leoCache
leo\core\leoCache.py:353:12: C2801: Unnecessarily calls dunder method __delitem__. Use del keyword. (unnecessary-dunder-call)
....************* Module leo.core.leoCommands
leo\core\leoCommands.py:4164:16: C2801: Unnecessarily calls dunder method __next__. Use next built-in function. (unnecessary-dunder-call)
..........************* Module leo.core.leoGlobals
leo\core\leoGlobals.py:398:15: C2801: Unnecessarily calls dunder method __hash__. Use hash built-in function. (unnecessary-dunder-call)
leo\core\leoGlobals.py:645:15: C2801: Unnecessarily calls dunder method __lt__. Use < operator. (unnecessary-dunder-call)
leo\core\leoGlobals.py:645:37: C2801: Unnecessarily calls dunder method __eq__. Use == operator. (unnecessary-dunder-call)
leo\core\leoGlobals.py:648:19: C2801: Unnecessarily calls dunder method __eq__. Use == operator. (unnecessary-dunder-call)
leo\core\leoGlobals.py:651:19: C2801: Unnecessarily calls dunder method __lt__. Use < operator. (unnecessary-dunder-call)
leo\core\leoGlobals.py:651:46: C2801: Unnecessarily calls dunder method __eq__. Use == operator. (unnecessary-dunder-call)
leo\core\leoGlobals.py:654:19: C2801: Unnecessarily calls dunder method __lt__. Use < operator. (unnecessary-dunder-call)
leo\core\leoGlobals.py:659:15: C2801: Unnecessarily calls dunder method __hash__. Use hash built-in function. (unnecessary-dunder-call)
leo\core\leoGlobals.py:2255:12: C2801: Unnecessarily calls dunder method __setattr__. Set attribute directly or use setattr built-in function. (unnecessary-dunder-call)
leo\core\leoGlobals.py:2299:12: C2801: Unnecessarily calls dunder method __setattr__. Set attribute directly or use setattr built-in function. (unnecessary-dunder-call)
........************* Module leo.core.leoNodes
leo\core\leoNodes.py:225:19: C2801: Unnecessarily calls dunder method __eq__. Use == operator. (unnecessary-dunder-call)
leo\core\leoNodes.py:239:15: C2801: Unnecessarily calls dunder method __eq__. Use == operator. (unnecessary-dunder-call)
leo\core\leoNodes.py:239:37: C2801: Unnecessarily calls dunder method __gt__. Use > operator. (unnecessary-dunder-call)
leo\core\leoNodes.py:242:15: C2801: Unnecessarily calls dunder method __eq__. Use == operator. (unnecessary-dunder-call)
leo\core\leoNodes.py:242:37: C2801: Unnecessarily calls dunder method __lt__. Use < operator. (unnecessary-dunder-call)
leo\core\leoNodes.py:245:19: C2801: Unnecessarily calls dunder method __eq__. Use == operator. (unnecessary-dunder-call)
leo\core\leoNodes.py:245:46: C2801: Unnecessarily calls dunder method __gt__. Use > operator. (unnecessary-dunder-call)
leo\core\leoNodes.py:1943:16: C2801: Unnecessarily calls dunder method __next__. Use next built-in function. (unnecessary-dunder-call)
...Exception on node <Attribute.connect l.288 at 0x1f0231fe1a0> in file 'C:\leo.repo\leo-editor\leo\core\leoPrinting.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.core.leoPrinting
leo\core\leoPrinting.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\core\leoPrinting.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-06.txt'. (fatal)
...............************* Module leo.core.leoserver
leo\core\leoserver.py:783:16: C2801: Unnecessarily calls dunder method __next__. Use next built-in function. (unnecessary-dunder-call)
..************* Module leo.external.leosax
leo\external\leosax.py:63:23: C2801: Unnecessarily calls dunder method __str__. Use str built-in function. (unnecessary-dunder-call)
..Exception on node <Attribute.connect l.67 at 0x1d0d133f490> in file 'C:\leo.repo\leo-editor\leo\plugins\QNCalendarWidget.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.QNCalendarWidget
leo\plugins\QNCalendarWidget.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\QNCalendarWidget.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.......Exception on node <Attribute.connect l.442 at 0x1194ed9a200> in file 'C:\leo.repo\leo-editor\leo\plugins\attrib_edit.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.attrib_edit
leo\plugins\attrib_edit.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\attrib_edit.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.....Exception on node <Attribute.connect l.122 at 0x272e973be80> in file 'C:\leo.repo\leo-editor\leo\plugins\bigdash.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.bigdash
leo\plugins\bigdash.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\bigdash.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.........................................................................Exception on node <Attribute.connect l.97 at 0x23362a698d0> in file 'C:\leo.repo\leo-editor\leo\plugins\livecode.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.livecode
leo\plugins\livecode.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\livecode.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.................Exception on node <Attribute.connect l.83 at 0x22d8f79bac0> in file 'C:\leo.repo\leo-editor\leo\plugins\nested_splitter.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.nested_splitter
leo\plugins\nested_splitter.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\nested_splitter.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
....Exception on node <Attribute.connect l.331 at 0x225d8ac71c0> in file 'C:\leo.repo\leo-editor\leo\plugins\nodetags.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.nodetags
leo\plugins\nodetags.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\nodetags.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.Exception on node <Attribute.connect l.207 at 0x12c6265ee00> in file 'C:\leo.repo\leo-editor\leo\plugins\nodewatch.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.nodewatch
leo\plugins\nodewatch.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\nodewatch.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
...........Exception on node <Attribute.connect l.94 at 0x1e87216d390> in file 'C:\leo.repo\leo-editor\leo\plugins\qt_commands.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.qt_commands
leo\plugins\qt_commands.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\qt_commands.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
..Exception on node <Attribute.connect l.689 at 0x1862ac6a8c0> in file 'C:\leo.repo\leo-editor\leo\plugins\qt_frame.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.qt_frame
leo\plugins\qt_frame.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\qt_frame.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.Exception on node <Attribute.connect l.363 at 0x21d92948b50> in file 'C:\leo.repo\leo-editor\leo\plugins\qt_gui.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.qt_gui
leo\plugins\qt_gui.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\qt_gui.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.Exception on node <Attribute.connect l.67 at 0x22eb67658a0> in file 'C:\leo.repo\leo-editor\leo\plugins\qt_idle_time.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.qt_idle_time
leo\plugins\qt_idle_time.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\qt_idle_time.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
....Exception on node <Attribute.connect l.500 at 0x1cb892eed40> in file 'C:\leo.repo\leo-editor\leo\plugins\qt_text.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.qt_text
leo\plugins\qt_text.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\qt_text.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
...Exception on node <Attribute.connect l.316 at 0x29da5907a90> in file 'C:\leo.repo\leo-editor\leo\plugins\quicksearch.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.quicksearch
leo\plugins\quicksearch.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\quicksearch.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
....Exception on node <Attribute.connect l.219 at 0x154151f3700> in file 'C:\leo.repo\leo-editor\leo\plugins\remove_duplicate_pictures.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.remove_duplicate_pictures
leo\plugins\remove_duplicate_pictures.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\remove_duplicate_pictures.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
..............Exception on node <Attribute.connect l.191 at 0x2b26fe09f00> in file 'C:\leo.repo\leo-editor\leo\plugins\stickynotes.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.stickynotes
leo\plugins\stickynotes.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\stickynotes.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.Exception on node <Attribute.connect l.625 at 0x1e00a5927d0> in file 'C:\leo.repo\leo-editor\leo\plugins\stickynotes_plus.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.stickynotes_plus
leo\plugins\stickynotes_plus.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\stickynotes_plus.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.....Exception on node <Attribute.connect l.30 at 0x229b8fe0fd0> in file 'C:\leo.repo\leo-editor\leo\plugins\threadutil.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.threadutil
leo\plugins\threadutil.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\threadutil.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.......Exception on node <Attribute.connect l.1664 at 0x1de7e4d70d0> in file 'C:\leo.repo\leo-editor\leo\plugins\viewrendered.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.viewrendered
leo\plugins\viewrendered.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\viewrendered.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.Exception on node <Attribute.connect l.2204 at 0x234ca9a2e30> in file 'C:\leo.repo\leo-editor\leo\plugins\viewrendered3.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.viewrendered3
leo\plugins\viewrendered3.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\viewrendered3.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
................Exception on node <Attribute.connect l.83 at 0x14e0d9417b0> in file 'C:\leo.repo\leo-editor\leo\plugins\xdb_pane.py'
Traceback (most recent call last):
  File "c:\leo.repo\ekr-pylint\pylint\utils\ast_walker.py", line 73, in walk
    callback(astroid)
  File "c:\leo.repo\ekr-pylint\pylint\checkers\typecheck.py", line 1057, in visit_attribute
    attr_parent = attr_node.parent
AttributeError: 'str' object has no attribute 'parent'
************* Module leo.plugins.xdb_pane
leo\plugins\xdb_pane.py:1:0: F0001: Fatal error while checking 'C:\leo.repo\leo-editor\leo\plugins\xdb_pane.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\Edward Ream\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-04-07-07.txt'. (fatal)
.....254 files, time: 1039.29 sec.
```

### Expected behavior

No pylint crashes :-) The pylint messages appear reasonable.

### Pylint version

```shell
pylint 2.14.0-dev0
astroid 2.11.2
Python 3.10.4 (tags/v3.10.4:9d38120, Mar 23 2022, 23:13:41) [MSC v.1929 64 bit (AMD64)]
```

### OS / Environment

Leo 6.6.1-devel, ekr-unit-tests branch, build 77f47d62b3
2022-04-06 13:50:52 -0500
Python 3.10.4, PyQt version 6.2.4
Windows 10 AMD64 (build 10.0.22000) SP0

### Additional dependencies

C:\leo.repo\leo-editor>python -m pip freeze

astroid==2.11.2
atomicwrites==1.4.0
attrs==21.4.0
beautifulsoup4==4.10.0
black==22.3.0
bleach==4.1.0
certifi==2021.10.8
charset-normalizer==2.0.12
click==8.1.0
colorama==0.4.4
coverage==6.3.2
cycler==0.11.0
defusedxml==0.7.1
dialite==0.5.3
dill==0.3.4
distlib==0.3.4
docutils==0.18.1
entrypoints==0.4
feedparser==6.0.8
filelock==3.6.0
flexx==0.8.3
fonttools==4.31.2
idna==3.3
ImageHash==4.2.1
iniconfig==1.1.1
isort==5.10.1
Jinja2==3.1.1
jsonschema==4.4.0
jupyter-client==7.2.1
jupyter-core==4.9.2
jupyterlab-pygments==0.1.2
kiwisolver==1.4.1
lazy-object-proxy==1.7.1
leo==6.6
lxml==4.8.0
Markdown==3.3.6
MarkupSafe==2.1.1
matplotlib==3.5.1
mccabe==0.7.0
mistune==0.8.4
mypy-extensions==0.4.3
nbclient==0.5.13
nbconvert==6.4.5
nbformat==5.2.0
nest-asyncio==1.5.5
numpy==1.22.3
packaging==21.3
pandocfilters==1.5.0
pathspec==0.9.0
Pillow==9.0.1
platformdirs==2.5.1
pluggy==1.0.0
pscript==0.7.7
py==1.11.0
pyenchant==3.2.2
Pygments==2.11.2
-e git+https://github.com/PyCQA/pylint.git@ba08c9d1ab8b624b4491a29bfc2f274dc00eeab8#egg=pylint
pyparsing==3.0.7
PyQt5==5.15.6
PyQt5-Qt5==5.15.2
PyQt5-sip==12.9.1
PyQt6==6.2.3
PyQt6-Qt6==6.2.4
PyQt6-sip==13.2.1
PyQt6-WebEngine==6.2.1
PyQt6-WebEngine-Qt6==6.2.4
pyrsistent==0.18.1
pytest==7.1.1
pytest-cov==3.0.0
python-dateutil==2.8.2
PyWavelets==1.3.0
pywin32==303
PyYAML==6.0
pyzmq==22.3.0
requests==2.27.1
scipy==1.8.0
Send2Trash==1.8.0
sgmllib3k==1.0.0
six==1.16.0
soupsieve==2.3.1
testpath==0.6.0
toml==0.10.2
tomli==2.0.1
tornado==6.1
tox==3.24.5
traitlets==5.1.1
types-docutils==0.18.0
types-requests==2.27.15
types-urllib3==1.26.11
typing_extensions==4.1.1
urllib3==1.26.9
virtualenv==20.14.0
webencodings==0.5.1
webruntime==0.5.8
websockets==10.2
windows-curses==2.3.0
wrapt==1.14.0

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #6221. Regression in 129c7302ceaabbbbf41a9872661a3550069b37e3 (unreleased 2.14)

- [ ] Add test

### CODE REVIEW

File: ChangeLog [hunk 1]
```diff
@@ -64,6 +64,11 @@ Release date: TBA
 
   Closes #1555
 
+* Fix crash (in unreleased development) in ``no-member`` check when inspecting C
+  extension modules with ``extension-pkg-allow-list``.
+
+  Closes #6221
+
```
Comment (Pierre-Sassoulas) [inline ChangeLog:71]:
```suggestion
```
I think we could add ``Closes #6221`` to the original log for the false negative (#1555 / #5544)

**Comment by jacobtylerwalls** [comment]
I spent a cycle trying to figure out how to test this. We would need a package with a C extension without ballooning our test dependencies. `coverage` is the package used for that already, but then we need an involved solution to reproduce the inference ambiguity on some element in `coverage`, possibly through inducing an import name collision via another locally installed package (see report, with `leo.core.Qt6` importing names from `Qt6` that might differ from the c extension `Qt6`. It's a bit overwhelming.

File: pylint/checkers/typecheck.py [hunk 2]
```diff
@@ -1053,6 +1053,10 @@ def visit_attribute(self, node: nodes.Attribute) -> None:
                 missingattr.add((owner, name))
                 continue
             else:
+                if isinstance(attr_nodes, nodes.NodeNG):
+                    # If owner was a living object, attr_nodes might be a node itself
+                    # Faster than retrieving the module object and testing _is_c_extension()
+                    break
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1059]:
```suggestion
                    break   # pragma: no cover
                    # In order to test this, we would need a package with a C extension without ballooning our test 
                    # dependencies. coverage is the package used for that already, but then we need an involved
                    # solution to reproduce the inference ambiguity on some element in coverage, possibly through 
                    # inducing an import name collision via another locally installed package (see report, with 
                    # leo.core.Qt6 importing names from Qt6 that might differ from the c extension Qt6.
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
I made a suggestion, we could also post the URL to this comment. I think adding the pragma no cover is important if we're voluntarily not testing the code.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
For future reference, here is a test case for `pylint` that fails if `PyQt6` is installed.
```diff
diff --git a/requirements_test.txt b/requirements_test.txt
index 2e35f584..04aac371 100644
--- a/requirements_test.txt
+++ b/requirements_test.txt
@@ -6,6 +6,7 @@ pre-commit~=2.17;python_full_version>="3.6.2"
 tbump~=6.6.0
 contributors-txt>=0.7.3
 pyenchant~=3.2
+PyQt6~=6.2
 pytest-cov~=3.0
 pytest-profiling~=1.7
 pytest-xdist~=2.5
diff --git a/tests/__init__.py b/tests/__init__.py
new file mode 100644
index 00000000..e69de29b
diff --git a/tests/functional/n/no/no_member_c_extension.py b/tests/functional/n/no/no_member_c_extension.py
new file mode 100644
index 00000000..e09f7d03
--- /dev/null
+++ b/tests/functional/n/no/no_member_c_extension.py
@@ -0,0 +1,9 @@
+"""Regression test for https://github.com/PyCQA/pylint/issues/6221"""
+
+# local import that creates ambiguity with test requirement PyQt6
+# PyQt6 has the name printsupport
+from tests.input.qt6 import printsupport
+
+dialog = printsupport.QPrintPreviewDialog
+# paintRequested (nodes.FunctionDef) is the owner of the name connect, not the module
+dialog.paintRequested.connect()
diff --git a/tests/functional/n/no/no_member_c_extension.rc b/tests/functional/n/no/no_member_c_extension.rc
new file mode 100644
index 00000000..b3320786
--- /dev/null
+++ b/tests/functional/n/no/no_member_c_extension.rc
@@ -0,0 +1,2 @@
+[MASTER]
+extension-pkg-allow-list=PyQt6.QtPrintSupport
diff --git a/tests/input/qt6.py b/tests/input/qt6.py
new file mode 100644
index 00000000..44312ea8
--- /dev/null
+++ b/tests/input/qt6.py
@@ -0,0 +1 @@
+from PyQt6 import QtPrintSupport as printsupport
```

**Comment by jacobtylerwalls** [comment]
Hold up, there's a deeper root cause that's fixable in the astroid brain for qt.

**Comment by jacobtylerwalls** [comment]
Superseded by https://github.com/PyCQA/astroid/pull/1505

**Comment by Pierre-Sassoulas** [comment]
But isn't it a good thing that we do not crash in case there's an issue in a brain or in astroid's analysis ?

**Comment by jacobtylerwalls** [comment]
I guess my thought was that we don't want to silence programming errors. Someone could contribute a broken brain and not realize it if we're just catching errors in pylint.

**Comment by Pierre-Sassoulas** [comment]
I agree we should not silence our own errors. But on the other hand, the brain was contributed with an error and we have a really hard time reproducing the bug, even adding an automated test to check the fix we added is complicated. What could make sure that we do not contribute faulty brains in the future ?

**Comment by jacobtylerwalls** [comment]
> Maybe better typing in astroid?

I hear that astroid is not ready for comprehensive type-checking. Maybe we can start turning it on file by file, or just in the brains and work our way toward the center. That's what I'm doing in another library I work on (start on edges).

I added some typing to the astroid PR that will raise mypy errors without the fix.

**Comment by Pierre-Sassoulas** [comment]
> I hear that astroid is not ready for comprehensive type-checking.

It's closer and closer though.

**Pull request closed without merging**